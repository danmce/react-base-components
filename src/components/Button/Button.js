import React from 'react'
import PropTypes from 'prop-types'
import ButtonStyles from './Button.styles'

const Button = props => <ButtonStyles {...props}>{props.children}</ButtonStyles>

Button.propTypes = {
  onClick: PropTypes.func,
  disabled: PropTypes.bool,
  type: PropTypes.oneOf([
    'primary',
    'secondary',
    'success',
    'error',
    'warning'
  ]),
  mode: PropTypes.oneOf(['filled', 'ghost', 'minimal', 'flat']),
  title: PropTypes.string,
  background: PropTypes.string,
  size: PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'xxl', 'xxxl']),
  stretched: PropTypes.bool,
  children: PropTypes.string
}

Button.defaultProps = {
  disabled: false,
  mode: 'filled',
  size: 'm',
  stretched: true,
  type: 'primary'
}

export { Button as default }
