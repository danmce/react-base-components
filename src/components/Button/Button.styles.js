import styled from '@emotion/styled'
import { darken } from '../../utils/colors'

const ButtonCommon = styled.button`
  align-items: center;
  appearance: none;
  background-color: ${props => backgroundSelector(props.theme, props.type)};
  border: ${props =>
    `${props.theme.borders.size.s} ${borderSelector(props.theme, props.type)}`};
  border-radius: ${props => props.theme.borders.radii.m}px;
  color: ${props => textColorSelector(props.theme, props.type)};
  cursor: pointer;
  display: flex;
  font-family: ${props => props.theme.typography.fontFamily};
  font-size: ${props => props.theme.typography.size.l}px;
  justify-content: center;
  outline: none;
  padding: 1rem 2rem;
  width: ${props => (props.stretched ? '100%' : 'auto')};
  &:hover {
    background-color: ${props =>
    backgroundSelector(props.theme, props.type, 'hover')};
    border: ${props =>
    `${props.theme.borders.size.s} ${borderSelector(
      props.theme,
      props.type,
      'hover'
    )}`};
  }
`

const backgroundSelector = (theme, type, mode) => {
  const color = theme.colors.button
  if (!mode) {
    switch (type) {
      case 'primary':
        return color.primary.background
      case 'secondary':
        return color.secondary.background
      case 'success':
        return theme.colors.success.fill
      case 'error':
        return theme.colors.error.fill
      case 'warning':
        return theme.colors.warning.fill
      default:
        return color.primary.background
    }
  } else if (mode === 'hover') {
    switch (type) {
      case 'primary':
        return darken(color.primary.background, 0.2)
      case 'secondary':
        return darken(color.secondary.background, 0.2)
      case 'success':
        return darken(theme.colors.success.fill, 0.2)
      case 'error':
        return darken(theme.colors.error.fill, 0.2)
      case 'warning':
        return darken(theme.colors.warning.fill, 0.2)
      default:
        return darken(color.primary.background, 0.2)
    }
  }
}

const borderSelector = (theme, type, mode) => {
  const color = theme.colors.button
  if (!mode) {
    switch (type) {
      case 'primary':
        return color.primary.border
      case 'secondary':
        return color.secondary.border
      case 'success':
        return theme.colors.success.fill
      case 'error':
        return theme.colors.error.fill
      case 'warning':
        return theme.colors.warning.fill
      default:
        return color.primary.border
    }
  } else if (mode === 'hover') {
    switch (type) {
      case 'primary':
        return darken(color.primary.border, 0.2)
      case 'secondary':
        return darken(color.secondary.border, 0.2)
      case 'success':
        return darken(theme.colors.success.fill, 0.2)
      case 'error':
        return darken(theme.colors.error.fill, 0.2)
      case 'warning':
        return darken(theme.colors.warning.fill, 0.2)
      default:
        return darken(color.primary.border, 0.2)
    }
  }
}

const textColorSelector = (theme, type) => {
  const color = theme.colors.button
  switch (type) {
    case 'primary':
      return color.primary.color
    case 'secondary':
      return color.secondary.color
    case 'success':
      return theme.colors.success.contrast
    case 'error':
      return theme.colors.error.contrast
    case 'warning':
      return theme.colors.warning.contrast
    default:
      return color.primary.color
  }
}

export default ButtonCommon
