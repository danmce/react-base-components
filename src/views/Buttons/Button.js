import styled from 'styled-components'

export default styled.button`
  ${props => props.theme.fontStyles.button};
  appearance: none;
  cursor: pointer;
  background-color: ${props => props.theme.colors.green};
  border: 0;
  border-radius: 1rem;
  padding: 2rem 4rem;
  outline: none;
`
