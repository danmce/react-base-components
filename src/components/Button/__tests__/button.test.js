import React from 'react'
import * as enzyme from 'enzyme'
import toJson from 'enzyme-to-json'
import Button from '../Button'
import UxPinWrapper from '../../../wrappers/UXPinWrapper'

test('renders error Button correctly', () => {
  const button = (
    <UxPinWrapper>
      <Button stretched={false} type='error' mode='filled' size='xs'>
        Test
      </Button>
    </UxPinWrapper>
  )

  expect(toJson(enzyme.mount(button))).toMatchSnapshot('enzyme.mount')
})
