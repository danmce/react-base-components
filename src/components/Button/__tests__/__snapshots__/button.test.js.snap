// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error Button correctly: enzyme.mount 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background-color: #B31612;
  border: 1 solid #B31612;
  border-radius: 3px;
  color: #FFFFFF;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 8px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  outline: none;
  padding: 1rem 2rem;
  width: auto;
}

.emotion-0:hover {
  background-color: rgb(143,18,14);
  border: 1 solid rgb(143,18,14);
}

<UXPinWrapper>
  <ThemeProvider
    theme={
      Object {
        "borders": Object {
          "radii": Object {
            "m": 3,
            "s": 0,
          },
          "size": Object {
            "s": "1 solid",
          },
        },
        "colors": Object {
          "button": Object {
            "primary": Object {
              "background": "#17384B",
              "border": "#2C546C",
              "color": "#FFFFFF",
            },
            "secondary": Object {
              "background": "transparent",
              "border": "#E9EEF2",
              "color": "#D7DFE2",
            },
          },
          "common": Object {
            "black": "#000000",
            "white": "#FFFFFF",
          },
          "error": Object {
            "contrast": "#FFFFFF",
            "fill": "#B31612",
          },
          "success": Object {
            "contrast": "#FFFFFF",
            "fill": "#00761D",
          },
          "warning": Object {
            "contrast": "#FFFFFF",
            "fill": "#E59600",
          },
        },
        "typography": Object {
          "family": "'Helvetica Neue', sans-serif",
          "size": Object {
            "l": 16,
            "m": 12,
            "s": 10,
            "xl": 20,
            "xs": 8,
            "xxl": 24,
            "xxxl": 34,
            "xxxxl": 50,
          },
          "weight": Object {
            "bold": 600,
            "light": 100,
            "regular": 300,
          },
        },
      }
    }
  >
    <Button
      disabled={false}
      mode="filled"
      size="xs"
      stretched={false}
      type="error"
    >
      <ButtonCommon
        disabled={false}
        mode="filled"
        size="xs"
        stretched={false}
        type="error"
      >
        <button
          className="emotion-0 emotion-1"
          disabled={false}
          mode="filled"
          size="xs"
          type="error"
        >
          Test
        </button>
      </ButtonCommon>
    </Button>
  </ThemeProvider>
</UXPinWrapper>
`;
